### PR TITLE
Starting to work on instructors notes

### PR DESCRIPTION
**Pull request by gvwilson**
This adds notes about how to teach Git - more contributions would be very welcome.

### CODE REVIEW

File: instructors.md [hunk 1]
```diff
@@ -3,28 +3,126 @@ layout: page
 title: Version Control with Git
 subtitle: Instructor's Guide
 ---
-## Legend
+Version control might be the most important topic we teach,
+but Git is definitely the most complicated tool.
```
Comment (jiffyclub) [inline instructors.md]: Reword this to make clear Git is the most complicated tool *we teach*. As written this means "Git is the most complicated tool *in existence*".

File: instructors.md [hunk 2]
```diff
@@ -3,28 +3,126 @@ layout: page
...
+However,
+GitHub presently dominates the open software forge landscape,
```
Comment (jiffyclub) [inline instructors.md]: I don't think young'uns are going to know what "forge" means in this context.

File: instructors.md [hunk 3]
```diff
@@ -3,28 +3,126 @@ layout: page
...
+    and,
+    "come on this really has to be the last version"
+    to motivate version control as a better way to collaborate
+    and as a better way to back work up.
 
-FIXME
+*   Version control is typically taught after the shell,
+    so collect learners' names during that session
+    and create a repository for them to share
+    with their names as both their IDs and their passwords.
```
Comment (jiffyclub) [inline instructors.md]: How do passwords come in here?

File: instructors.md [hunk 4]
```diff
@@ -3,28 +3,126 @@ layout: page
...
 
-## [Conflicts](03-conflict.html)
+*   If your learners are advanced enough to be comfortable SSH,
```
Comment (jiffyclub) [inline instructors.md]: "comfortable *with* SSH"

File: instructors.md [hunk 5]
```diff
@@ -3,28 +3,126 @@ layout: page
...
+    tell them they can use keys to authenticate on GitHub instead of passwords,
+    but don't try to set this up during class:
+    it takes too long,
+    and is a distraction from the core ideas of the lesson.
 
-FIXME
+*   Be very clear what files learners are to edit
+    and what user IDs they are to use
+    when giving instructions:
```
Comment (jiffyclub) [inline instructors.md]: Move "When giving instructions" to the beginning of the sentence so it's clear it is directed at the reader.

**Comment by jiffyclub** [comment]
From the instructor notes it sounds like the instructor and students are all working in the same repository, but that isn't stated explicitly anywhere. I also rarely see that situation in practice, I've mostly seen teachers have the students work in pairs in each others' repos.

File: instructors.md (hunk 1 above)
Comment (gvwilson) [inline instructors.md]: > Reword this to make clear Git is the most complicated tool /we teach/. 
> As written this means "Git is the most complicated tool /in existence/".
>
There are days...

**Comment by BillMills** [comment]
I don't usually teach `diff`. I find this confuses students (particularly on what it's comparing before and after an `add`), and you can get most of the same mileage out of the nice red and green highlights in your github repo - which in all honesty are all I've ever needed in my personal practice.

Instead, I like to focus on a simple workflow - 'pull add commit push' - which will get most users through most days on most small projects - and spend some extra time on merge conflicts, so they can recover from problems as they arise.

**Comment by andreww** [comment]
One thing that caused confusion for some in our recent workshops comes from "recovering old versions" towards the end of "a better kind of backup". If, instead of doing `$ git checkout f22b25e mars.txt`, you (or a learner) does `$ git checkout f22b25e` you get  the "in 'detached HEAD' state" message and confusion abounds.  It's then possible to keep on committing, but things like `git push origin master` a bit later will not give easily comprehensible results. It also makes it look like you can loose commits. It's probably worth explicitly warning about this somewhere in these notes.

**Comment by andreww** [comment]
Is it worth including some information on the expected background of the learners for these lessons? For example, I imagine they they won't have experience of any version control system (which may be an advantage when it comes to learning git). This may make it easer to select between these and a possible future "git-intermediate" set, and help make the rational for the design decisions more obvious.

**Comment by gvwilson** [comment]
There's a `Prerequisites` section in `index.md` - a PR against that 
would be very welcome.

File: instructors.md (hunk 2 above)
Comment (jni) [inline instructors.md]: Even old'uns don't know it! What does it mean? LOL

**Comment by BillMills** [comment]
@andreww I saw a brilliant introduction to git recently, that referenced version control background, from Sam White:
"Who knows what version control is?"
*nobody*
"Who has ever used an undo feature before?"
*everyone*
"Undo is the simplest type of Version Control."
GASP

In other words - the novice version control lesson is just that, a first introduction, no prerequisites. And yet, there is an opportunity to show our students this is not so alien, after all.

File: instructors.md [hunk 6]
```diff
@@ -3,28 +3,126 @@ layout: page
...
-## Legend
+Version control might be the most important topic we teach,
+but Git is definitely the most complicated tool.
+However,
+GitHub presently dominates the open software forge landscape,
+so we have to help novices learn just enough Git
+to feel they can and should learn more on their own.
 
-Wolfman and Dracula will work together in one project.
+This is why we don't teach branching:
```
Comment (jni) [inline instructors.md]: How long is the workshop? I've been (I think) quite successful in teaching branching and PRs in a 4h session... I do make *some* UNIX and one programming language a prerequisite though...

File: instructors.md (hunk 2 above)
Comment (mikej888) [inline instructors.md]: Rephrase "However, GitHub presently dominates..." to "However, the project hosting service GitHub, which provides Git as one of its services, presently dominates..."
A number of students think Git and GitHub are the same or one needs GitHub to use Git.

File: instructors.html [hunk 7]
```diff
@@ -27,21 +27,45 @@
         <div class="span10 offset1">
           <h1 class="title">Version Control with Git</h1>
           <h2 class="subtitle">Instructor's Guide</h2>
-<h2 id="legend">Legend</h2>
-<p>Wolfman and Dracula will work together in one project.</p>
-<h2 id="overall">Overall</h2>
+<p>Version control might be the most important topic we teach, but Git is definitely the most complicated tool. However, GitHub presently dominates the open software forge landscape, so we have to help novices learn just enough Git to feel they can and should learn more on their own.</p>
```
Comment (mikej888) [inline instructors.html]: A number of students think Git and GitHub are the same or one needs GitHub to use Git.
So, rephrase "However, GitHub presently dominates..." to "However, the project hosting service GitHub, which provides Git as one of its services, presently dominates..."
Or just defer mentioning GitHub till later?

Comment (swaldman3) [inline instructors.html]: Agreed that this paragraph comes close to conflating git and github. It's also true that git presently dominates the open software landscape, so why not just say that?

File: instructors.md (hunk 6 above)
Comment (r-gaia-cs) [inline instructors.md]: > How long is the workshop?

The Git section normally take one morning or one afternoon.

> I've been (I think) quite successful in teaching branching and PRs in a 4h session...

Depending of how much challenges do you leave the students do and how many helpers do you have will be possible to teach branching but we mostly agree that this is a intermediate feature so we prefer not teaching it for novices.

File: instructors.md [hunk 8]
```diff
@@ -3,28 +3,126 @@ layout: page
...
+
+### Gedit
+
+You should use
+
+~~~ {.bash}
+$ git config --global core.editor 'gedit --standalone'
+~~~
+
+to avoid this error occurring
```
Comment (r-gaia-cs) [inline instructors.md]: Maybe

    to avoid

    ~~~~ {.bash}
    $ git commit
    ~~~
    ~~~ {.error}
    Aborting commit due to empty commit message.
    ~~~

    that occurs

**Comment by abostroem** [comment]
@jiffyclub I agree about the 1 vs many repositories. I find when all students are working in the same repo it leads to too many conflicts and too many people trying to push at the same time. Some instructors solve this by having one person demonstrate for the class but I prefer to give everyone hands on experience (I think it makes it less scary if you've done it at least once). My preference is to have everyone create a remote repository - learn how to connect it to their local repository. Then put them in pairs have one person clone the other's repository and then push and pull with varying levels of complexity.

**Comment by abostroem** [comment]
@BillMills I like to teach diff during the local git section. This gives them a tools set to track their work even if they never want to collaborate (with themselves on another machine or someone else) and off line.

**Comment by abostroem** [comment]
@jni I used to teach branching in a 4 hour session. Here are the trade offs in my opinion:
Pro:
* You can set up local conflicts so that students learn how to resolve them before adding the complication of a remote repository and a second person.
* Merging in the remote section makes more sense
* Its good to know about if you have time

Con:
It adds 30 minutes to an hour to the lesson

File: instructors.html (hunk 7 above)
Comment (willingc) [inline instructors.html]: I think it is important to let people know that there is a distinction between Git and GitHub. Although GitHub is a popular service, the availability of more private repos on other services warrants making a distinction between Git the tool and hosting services. I feel pretty strongly that if one mentions GitHub that other hosting service should be mentioned.

Possible text:
Git is a flexible and fast version control tool. Users have the option to working with the tool on just their computer or to use the tool in a distributed network of other remote computers and servers.

Many users use a cloud hosted version control service, like GitHub, for ease of collaboration and sharing of source code. While GitHub is currently the most popular version control service, Bitbucket, GitLab, and other services also provide hosting of repositories.

Many users like the convenience and collaboration that

File: instructors.html [hunk 9]
```diff
@@ -27,21 +27,45 @@
...
+<p>This is why we don't teach branching: while it is a power tool in the hands of a knowledgeable user, it is an extra cognitive burden for someone who is new to the idea of version control. This is also why we don't get into hashes and commit objects with novices, but try instead to convince them that version control is:</p>
```
Comment (willingc) [inline instructors.html]: Suggested rewording:
We don't introduce branching in the novice lesson. While branching extends the power and flexibility of collaborating with others, branching concepts and the variety of possible workflows adds too much cognitive burden for someone who is new to the idea of version control. For similar reasons, we don't get into hashes and commit objects with novices, but try instead to convince them that version control is:

File: instructors.html [hunk 10]
```diff
@@ -27,21 +27,45 @@
...
+<li><p>If some learners are using Windows, there will inevitably be issues merging files with different line endings. (Even if everyone's on some flavor of Unix, different editors may or may not add a newline to the last line of a file.) Take a moment to explain these issues, since learners will almost certainly trip over them again.</p></li>
+<li><p>We don't use a Git GUI in these notes because we haven't found one that installs easily and runs reliably on the three major operating systems, and because we want learners to understand what commands are being run. That said, instructors should demo a GUI on their desktop at some point during this lesson and point learners at <a href="http://git-scm.com/downloads/guis">this page</a>.</p></li>
+<li><p>Instructors should also show learners graphical diff/merge tools like <a href="https://sourcegear.com/diffmerge/">DiffMerge</a>.</p></li>
+<li><p>When appropriate, explain that we teach Git rather than CVS, Subversion, or Mercurial primarily because of GitHub's growing popularity: CVS and Subversion are now seen as legacy systems, and Mercurial isn't nearly as widely used in the sciences right now.</p></li>
```
Comment (willingc) [inline instructors.html]: It would be better to mention that Git is faster and a better collaboration tool than CVS or Subversion. Popular services such as GitHub, Bitbucket or GitLab also make using Git an excellent tool choice. Mercurial, though similar to Git, has a smaller user base and many of the Git concepts apply if a user has a future need for Mercurial.

File: instructors.md (hunk 2 above)
Comment (willingc) [inline instructors.md]: Agree. Strike the forge.

File: instructors.md (hunk 6 above)
Comment (willingc) [inline instructors.md]: There are excellent branching tutorials and simulations available on-line. If you teach branching, you run a high likelihood of confusing those new to version control and/or command line use. Also branching adds complexity since there are various workflows that can be used in practice. Better to make it an intermediate lesson or advanced lesson (even if self directed for those that already know git).

File: instructors.md (hunk 3 above)
Comment (willingc) [inline instructors.md]: Perhaps passwords come in if a user has to create a new GitHub account, the user should be instructed to remember the user name and password for the GitHub account since they will need it again. (Sounds obvious yet more students than one would expect forget their password). I don't think there is any reason an instructor would need a user's password.

File: instructors.html [hunk 11]
```diff
@@ -27,21 +27,45 @@
...
+<li><p>Version control is typically taught after the shell, so collect learners' names during that session and create a repository for them to share with their names as both their IDs and their passwords.</p></li>
+<li><p>If your learners are advanced enough to be comfortable SSH, tell them they can use keys to authenticate on GitHub instead of passwords, but don't try to set this up during class: it takes too long, and is a distraction from the core ideas of the lesson.</p></li>
```
Comment (jhamrick) [inline instructors.html]: "to be comfortable with SSH"

**Comment by jhamrick** [comment]
Sorry for being late to the party (I've been swallowed by a paper deadline the last couple of weeks). I think this looks great! Mostly, I agree with comments others have had. I'm also a little confused about where passwords come in, too -- in my experience, it's worked well to either just have people create local git repositories on their own machines with `git init` (then they know how to create them in the future, too), or to clone stuff from GitHub and push/pull from that.
